src/adapter/src/rbac.rs

```diff
@@ -137,6 +137,47 @@ pub fn check_command(catalog: &Catalog, cmd: &Command) -> Result<(), Unauthorize
     }
 }
 
+/// Checks if a session is authorized to purify a CREATE SOURCE statement. Usually authorization is
+/// checked after planning by [`check_plan`]. However CREATE SOURCE statements are purified
+/// before planning, which may require the use of some connections and secrets.
+pub fn check_purify_create_source(
+    catalog: &impl SessionCatalog,
+    session: &Session,
+    resolved_ids: &ResolvedIds,
+) -> Result<(), AdapterError> {
+    let current_role_id = session.current_role_id();
```

Comment:
The various checks here are duplicated in `create_plan`. Could you move both sets to a common function so we don't miss updating one if we need to change the common code?

Reply:
Yeah, it'll probably have to be a macro because of the instances of `return Ok(())`.